rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,757 @@
+
+                if "dense_dropout" in self._tf_layers:
+                    f = self._tf_layers["dense_dropout"](f, training)
+
+            dense_features.append(f)
+
+        return tf.concat(dense_features, axis=-1)
+
+
+class ConcatenateSequenceSentenceFeatures(tf.keras.layers.Layer):
+    """Combines two tensors (sentence-level and sequence level features) into one.
+
+    This layer concatenates sentence- and sequence-level (token-level) features (each
+    feature type represented by a single dense tensor) into a single feature tensor.
+    When expanded to the same shape, sentence-level features can be viewed as sequence-
+    level ones, but with the sequence length being 1 for all examples. Hence, features
+    of the two types can be concatenated along the sequence (token) dimension,
+    effectively increasing the sequence length of each example by 1. Because sequence-
+    level features are all padded to the same max. sequence length, the concatenation is
+    done in the following steps:
+    1. Optional: If the last dimension size of sequence- and sentence-level feature
+        differs, it's unified by applying a dense layer to each feature types.
+    1. sentence-level features are masked out everywhere except at the first available
+        position (`sequence_length+1`, where the length varies across input examples)
+    2. the padding of sequence-level features is increased by 1 to prepare empty space
+        for sentence-level features
+    3. sentence-level features are added just after the sequence-level ones
+
+    Arguments:
+        attribute: Name of attribute (e.g. `text` or `label`) whose features will be
+            processed.
+        concat_dimension: A positive integer -- number of output units of the dense
+            layers used to unify the last dimension of each feature type.
+        feature_signature_sequence: Feature signature describing the sequence-level features.
+        feature_signature_sentence: Feature signature describing the sentence-level features.
+        config: A model config for correctly parametrising the layer's components.
+
+    Input shape:
+        Tuple of three 3-D dense tensors, with shapes:
+            sequence: `(batch_size, max_seq_length, input_dim_seq)`
+            sentence: `(batch_size, 1, input_dim_sent)`
+            mask: `(batch_size, max_seq_length+1, 1)`
+
+    Output shape:
+        3-D tensor with shape: `(batch_size, sequence_length, output_dim)` where:
+        - `output_dim` matches `input_dim_seq` if that is the same as `input_dim_sent`,
+            otherwise `output_dim` = `concat_dimension`.
+        - `sequence_length` is the sum of 2nd dimension sizes of arguments `sequence`
+            and`sentence`.
+    """
+
+    def __init__(
+        self,
+        attribute: Text,
+        concat_dimension: int,
+        feature_signature_sequence: FeatureSignature,
+        feature_signature_sentence: FeatureSignature,
```

Comment:
it seems to me that you only use `units`, let's pass only `sequence_units` and `sentence_units`, passing the whole signature creates a false sense of complexity

Reply:
I kinda disagree. Passing the (lightweight) signatures creates consistency. Also, this line:
```python
if feature_signature_sequence and feature_signature_sentence:
```
will look weird with units instead of signatures... Would you say that we passed (or have as a default) the value of `0` (or `-1`?) when there are no features for the feature type?

Reply:
> signatures creates consistency

I don't think so. All layers are different. Also, because of this, you need to create fake signatures in the other layer. But most importantly it hides what the layer is doing

let's follow Occam's razor principle

Reply:
> Would you say that we passed (or have as a default) the value of 0 (or -1?) when there are no features for the feature type?

why not pass `None`?

Reply:
Alright, you've convinced me 🙂